Sib 101

JakubMakarewiczCap opened the pull request:
Implementation of api communicators

JakubMakarewiczCap committed:
Some initial classes

JakubMakarewiczCap committed:
eod commit, will refactor tomorrow
interface methods, api client

JakubMakarewiczCap committed:
refactoring, added ApiHolder

JakubMakarewiczCap committed:
invoke->invokeGet

JakubMakarewiczCap committed:
json parser

JakubMakarewiczCap committed:
Added response code handling to ApiClient
Removed unused methods, renamed classes

JakubMakarewiczCap committed:
parsing json in ApiClient.java

JakubMakarewiczCap committed:
code formatting

JakubMakarewiczCap committed:
added constructors to DTOs

JakubMakarewiczCap committed:
reorganized RequestBuilder and ApiClient

JakubMakarewiczCap committed:
initial implementation of CMC communicator

JakubMakarewiczCap committed:
implementation of CMC
implemented getCurrentListing for CoinMarketCap

JakubMakarewiczCap committed:
implementation of ApiHolder
implemented getCoinMarketData

JakubMakarewiczCap committed:
fix to previous commit

JakubMakarewiczCap committed:
implementation of CMC
implemented historical quote request

JakubMakarewiczCap committed:
implementation of ApiHolder
implemented historical data, fixed return types

JakubMakarewiczCap committed:
Hotfix

JakubMakarewiczCap committed:
DTO changes

JakubMakarewiczCap committed:
Testing
Written Unit Tests for most methods in .apiCommunicator

JakubMakarewiczCap committed:
CoinMarketCap implementation
switched to production api

JakubMakarewiczCap committed:
Testing
fixed client urls, new unit tests

JakubMakarewiczCap committed:
Testing
added useless DTO tests

JakubMakarewiczCap committed:
Bug Fixing
fixed bugged nominal delta calculation in CMCParser

JakubMakarewiczCap committed:
Testing
Integration test for cmcCommunicator, unit test for parser.calculate nominal delta

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/ApiHolder.java:
```diff
@@ -0,0 +1,74 @@
+        add(coinName);
+      }}, new ArrayList<>() {{
+        add("usd");
+      }});
```
Hah, it's the very same thing I did as a junior in one of first tasks. Such convenient way to instantiate one-element ArrayList has quite significant side-effect: both definitions above are separate anonymous classes that are defined solely for this purpose. I wouldn't like to find out how serious are implications of writing a large solution that is filled with such ways to instantiate small arrays in many places.

Look at discussion here (but take into account that this discussion compares valid use case to define a class one way or another - in your case it is not needed to define a new class):
https://stackoverflow.com/questions/3028341/java-anonymous-class-efficiency-implications

There are convenient oneliners to instantiate `ArrayList`, but I'll leave it to you to discover them. Be cautious about StackOverflow - it may contain decade-old answers that don't take into account conveniences introduces somewere along Java 7 - Java 17.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/ApiHolder.java:
```diff
@@ -0,0 +1,74 @@
+    Iterator<Entry<ApiProviderEnum, ApiCommunicator>> iterator = this.apiCommunicators.entrySet()
+        .iterator();
+    boolean stop = false;
+    while (iterator.hasNext() && !stop) {
```
Several comments to this loop:
- I'm not sure if we want to return all results including 'history' of failures until successful result. It may make more sense to log failures for use by developers and application support, but keep this information away from end user or even from other layers of application. But keep it if there is a valid case to return this information.
- It can be done in more elegant way using Streams API - regardless if you will return all attempts until success or only success,
- Even without Streams API, you can use for-each loop with return statement inside loop. It's very questionable though - many developers consider it a rule to have only one return statement at the bottom of method. In my personal opinion it doesn't matter a lot if you write clean code that makes such return statement one of few statements in small, simple method that is created for single responsibility like collecting all results until success. What matters most is that another developer who reads your code will immediately know what your method does.
- In case if we would decide to use loop in this shape anyways, I'd rather rename `stop` to something more descriptive - at least `operationSuccessful` or even something that would be meaningful in this context.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/ApiHolder.java:
```diff
@@ -0,0 +1,74 @@
+    }
+    return results;
+  }
+}
```
Method above is largely a duplicate of previous method - only thing that differs is call on `ApiCommunicator`. You can improve it with functional interface or you can try other design call to avoid such duplication.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinGeckoCommunicator/CoinGeckoCommunicator.java:
```diff
@@ -0,0 +1,31 @@
+@Component
+public class CoinGeckoCommunicator implements ApiCommunicator {
+
+  ApiClient client = new ApiClient();
```
How would you mock `ApiClient` in tests? Wouldn't it be better to use dependency injection to instantiate `ApiClient` outside of this class and have more flexibility to write actual unit tests?

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapClient.java:
```diff
@@ -0,0 +1,59 @@
+  private final String version = "v2";
+  private final String url = path + "/" + version; // TODO: move to env
+
+  private final String key = "dc8ad4b3-5f8a-4190-84e9-34efc1fd81a9";
```
All 4 variables above shouldn't be hardcoded - `application.properties` are completely empty so far on your branch.

Commit:
Coingecko implementation.
Initial coingecko implementation for current and historical data with basic tests.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapCommunicator.java:
```diff
@@ -0,0 +1,69 @@
+  private static final Logger log = LogManager.getLogger(CoinMarketCapCommunicator.class);
+
+  @Override
+  public ApiProviderEnum getApiProvider() {
```
Naming of classes, methods, fields is unnecessarily confusing and inconsistent in whole context of strategy pattern implementation here. On top of that, this enum is not `ApiProvider`. Actually, `ApiHolder` has this role. I think that `ApiProviderEnum` is rather `ApiCommunicatorStrategy`.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapClient.java:
```diff
@@ -0,0 +1,59 @@
+      long timestamp) throws IOException {
+    LinkedHashMap<String, String> queryParams = new LinkedHashMap<>() {{
+      put("symbol", String.join(",", coins));
+      put("convert", String.join(",", vsCurrencies));
```
You reuse parameters like `"symbol"` or `"convert"` in many places. 
- It is helpful to have them extracted as constants or enum, for example to see that this `"count"` is part of specific API or to have one place which will tell you what other API possibilities you have (such enum may be a good place to put a link to API documentation for reference purpose). 
- It also helps if some parameter gets renamed (for example because it is your API and you want to rename parameter to more descriptive name). 
- What's more, maybe it will make sense to define a request object that will ensure a typesafe API, for example enforcing `"count"` parameter to be `int`.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapCommunicator.java:
```diff
@@ -0,0 +1,69 @@
+      if (ret.getStatus() == ResultStatus.SUCCESS &&
+          ret.getCoinMarketDataDTOS().size() != coins.size()) {
+        ret.setStatus(ResultStatus.PARTIAL_SUCCESS);
+      }
```
This `if` condition looks clearly as a candidate to put it inside `parseQuiteLatestResult` method, given that it will be fed with information about `coins.size()`.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapCommunicator.java:
```diff
@@ -0,0 +1,69 @@
+      return new CoinMarketDataResult(this.providerEnum, ResultStatus.FAILURE, e.getMessage(),
+          null);
+    }
+  }
```
Above is also largely a duplicate of previous method. Common denominator of both is that you have two versions of similar code - one to get latest result, another to get historical result. Maybe you could thing of a way to make abstract code that would take one or another action (strategy) based on provided request parameters? Maybe some layers can be designed in a way that they will be completely oblivious to the fact if they are processing latest or historical request.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+
+  private static final Logger log = LogManager.getLogger(CoinMarketCapResponseParser.class);
+
+  public String parseStatus(JsonNode status) {
```
That's rather `parseError`. BTW is it impossible that service could return null `"error_message"`? If API gives any chance for it to be the case, then this code may cause failure someday.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+      result.setStatus(ResultStatus.FAILURE);
+      result.setErrorMessage(e.getMessage());
+      return result;
+    }
```
Corresponding to my other comment, this code actually is a great candidate to have single return statement at the end of method to indicate a contract here.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+
+      result.setStatus(ResultStatus.SUCCESS);
+      result.setCoinMarketDataDTOS(
+          this.parseCoinsQuoteLatestResult(apiResponseBody.get("data")));
```
These two lines above are part of `else` statement. It explicitly indicates that you are doing something in scenario opposite to the one indicated in `if` condition. Any code below `if` block should be unrelated to the `if` condition. This way you clearly indicate the flow.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+          add(new DeltaDto(IntervalEnum.ONE_DAY, pctChange24h, nominalChange24h));
+          add(new DeltaDto(IntervalEnum.SEVEN_DAY, pctChange7d, nominalChange7d));
+          add(new DeltaDto(IntervalEnum.THIRTY_DAY, pctChange30d, nominalChange30d));
+        }};
```
Duplicated (or _quadruplicated_) code above. Please extract method that will create `DeltaDto` based on `quote` and `IntervalEnum` parameters.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  public ArrayList<CoinMarketDataDto> parseCoinsQuoteLatestResult(JsonNode data)
```
This is **very large** method. Ideally, a single method should consist of few lines of code and should call other methods (possibly in different classes) to drill down into details of some operations.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+      for (Map.Entry<String, ObjectNode> quote : quotesConverted.entrySet()) {
+        var currentPrice = quote.getValue().get("price").asDouble();
+        var timestamp = TimeFormatter.convertStringToTimestamp(
+                quote.getValue().get("last_updated").asText(), "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
```
Consider overloading `convertStringToTimestamp` to have default pattern inside - it is the only pattern you use in whole project so far.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+      result.setErrorMessage(e.getMessage());
+      return result;
+    }
+  }
```
Above is another example of duplicate between latest and historical variant that differs only at one method called inside.

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+        var currentPrice = quote.getValue().get("price").asDouble();
+        var timestamp = TimeFormatter.convertStringToTimestamp(
+                quote.getValue().get("last_updated").asText(), "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+            .getTime();
```
Also, you never use `Timestamp` in any other way than calling `getTime()` on it in the same line. What's the point of returning `Timestamp` instead of `long` then?

piotr-j-cg commented on src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/coinMarketCapCommunicator/CoinMarketCapResponseParser.java:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  public ArrayList<CoinMarketDataDto> parseCoinsQuoteHistoricalResult(JsonNode data)
+      throws ParseException {
```
This method consists too much different responsibilities. In quick refactor I have extracted a bunch of methods that made hierarchy of actions much more visible:
- `extractCoinMarketDataDtos`
- `extractCoinMarketDataDto`
- `extractQuoteDtos`
- `extractQuoteDto`
- `convertNodeToMap`
I don't mention it as preferable division of code, I just did it to make it more readable for me (so I don't share the result). Please consider splitting this code into multiple parsers that create respective objects. Or some other solution to make this code more readable.

piotr-j-cg reviewed (commented):
Consider comments on this review to be similar to what you would receive from me in real life scenario. Some comments may raise the bar quite high for you considering strict amount of time that you have for your task. It may take a lot of trials and failures to achieve code practice which would let you tackle different scenarios. So don't consider my comments as requirements to fix your code, but rather suggestions of areas which you may improve if you will have time for it.

On positive side, I saw a lot of code where I would wish that it would be written this way.

On potential room for improvement, I think it is most important for you to work on a code that consists of small classes (target 1 class = max 1 screen) with small methods (even 2-5 statements per method wherever it makes sense). In case of data classes it is achieved, but business logic is accumulated around few classes and sometimes in large methods.

And something that is considered as most difficult task by many: naming - please pay attention to make it consistent and descriptive (to the point - make it deliver the purpose and information about domain, avoid ambiguous words like 'handler').

piotr-j-cg commented on src/test/java/com/capgemini/fs/apiCommunicator/coinGeckoCommunicator/CoinGeckoCommunicatorTest.java:
```diff
@@ -0,0 +1,31 @@
+    ArrayList<String> vsCurrencies = new ArrayList<>();
+    vsCurrencies.add("usd");
+    //vsCurrencies.add("eur");
+    s.getCurrentListing(coins, vsCurrencies);
```
There is no single assertion or Mockito verification in this test class. Without them, we actually don't check returned values and we don't verify flow of application by verifying what was called on mocks (if applicable).

JakubMakarewiczCap committed:
Bug Fixing
Small fixes regarding dependency injection

JakubMakarewiczCap committed:
Bug Fixing
Small fixes

JakubMakarewiczCap committed:
Testing
ApiHolder initial test suite

JakubMakarewiczCap committed:
Bug fixing
Final bug fixing

JakubMakarewiczCap committed:
Refactoring
Builder pattern initial class and interface

JakubMakarewiczCap committed:
Reverted CoinDashboardApplication changes
